lib/inventory/inventory.go

```diff
@@ -260,30 +272,191 @@ type UpstreamHandle interface {
 	// Hello gets the cached upstream hello that was used to initialize the stream.
 	Hello() proto.UpstreamInventoryHello
 
-	Ping(ctx context.Context) (d time.Duration, err error)
+	Ping(ctx context.Context, id uint64) (d time.Duration, err error)
 	// HasService is a helper for checking if a given service is associated with this
 	// stream.
 	HasService(types.SystemRole) bool
+
+	// VisitInstanceState runs the provided closure against a representation of the most
+	// recently observed instance state, plus any pending control log entries. The returned
+	// value may optionally include additional control log entries to add to the pending
+	// queues. Inputs and outputs are deep copied to avoid concurrency issues. See the InstanceStateTracker
+	// for an explanation of how this system works.
+	VisitInstanceState(func(ref InstanceStateRef) InstanceStateUpdate)
+
+	// HeartbeatInstance triggers an early instance heartbeat. This function does not
+	// wait for the instance heartbeat to actually be completed, so calling this and then
+	// immediately locking the instanceStateTracker will likely result in observing the
+	// pre-heartbeat state.
+	HeartbeatInstance()
+}
+
+// instanceStateTracker tracks the state of a connected instance from the point of view of
+// the inventory controller. Values in this struct tend to be lazily updated/consumed. For
+// example, the LastHeartbeat value is nil until the first attempt to heartbeat the instance
+// has been made *for this TCP connection*, meaning that you can't distinguish between an instance
+// that just joined the cluster from one that just reconnected by looking at this struct. Similarly,
+// when using this struct to inject control log entries, said entries won't be included until the next
+// normal instance heartbeat (though this may be triggered early via UpstreamHandle.HeartbeatInstance()).
+// The primary intended usage pattern is for periodic operations to append entries to the QualifiedPendingControlLog,
+// and then observe wether or not said entries end up being included on subsequent iterations. This
+// patterns lets us achieve a kind of lazy "locking", whereby complex coordination can occur without
+// large spikes in backend load. See the QualifiedPendingControlLog field for an example of this pattern.
+type instanceStateTracker struct {
+	// MU protects all below fields and must be locked during access of any of the
+	// below fields.
+	MU sync.Mutex
```

Comment:
It's a little strange to export the mutex but not all the fields it protects. Nothing to stop other packages from locking/unlocking the mutex at the wrong time.

How hard would it be to unexport the mutex and have a concurrency-safe API for clients?